### ISSUE TITLE
Sharing information - Printing Logs and Saving States

### ISSUE THREAD

**Issue by kodomazer** [evolvio/evolv.io#82]
We need to finish filling out the methods for saving a text log of the situation.  

We should also try and get saving and loading of the board state as well.  I can work on saving and loading Brains, but it would be nice to hear about how people think we should go about doing that.  E.G. saving each brain in an individual file vs saving it as one big file and saving in plain text vs saving as a binary.

Lower priority would be to try and make the images simply the board and not all of the UI elements that are on the side.

I'll put in the stitched together version of the latest run that I made with the new Brains over here once I figure out how to do that...

**Comment by 10carfernandez** [comment]
I was just going to start doing that yesterday but decided to rebase to this version first since I've already made some unfinished changes that aren't really bug fixes. 

Here is what I was thinking as far as how I was planning on saving brains/creatures:

- Save as a text file

- Save all brains/creatures into one file

- Save brain/creatures in the current view. E.g., if you're zoomed into just 3 creatures and you hit "save," only those three creatures will be saved to the one text file; if you hit "save" with the zoom reset, then it will save all creatures on the board.

- Save highlighted creatures. Right now, we can only select one creature. But if we could implement a method for highlighting them similar to how you highlight icons on a desktop, it would be more useful. Highlighted creatures would take precedence over the previous scenarios. So if you've highlighted some creatures and then zoomed in, the ones not in the view but still highlighted would still be saved. 

I think board states could be done after implementing a method for saving creatures. So when creatures are saved, a set of coordinates (ie, where they were saved) and where they were pointed at could be included for each of them, and the seed they were all under. Then, the creatures could be reloaded to those exact coordinates once a load method for the board state is created.

**Comment by kodomazer** [comment]
We've also been discussing keeping track of lineages for various reasons.

In #73 I bring up the idea of remembering "successful" creatures, we might want a way to automatically identify and save very successful creatures.

In #51  there is talk about building a family tree.  The question relevant to here would be, how are we going to save that information?  As one file that gets updated as the tree gets extended?  As a set of files that all point to each other?

In #64 Speciation could be related to a common ancestor, if you don't share the ancestor within the past 5-10 generations then you're a separate species or something like that.  Or we could have softer barriers for speciaction, like Brain structure (with the new system I'm working on)

**Comment by kodomazer** [comment]
It's too big to upload to github, so here's a gfycat link.
https://gfycat.com/UglyLavishHyracotherium

**Comment by 10carfernandez** [comment]
Yeah, lineages would definitely be good. 

Here's some info that might help for #73: In biology, the fitness of an organism (ie, how "successful" it is) is defined as the product of fecundity and survivability. Fecundity is the expected number of offspring during the organism's lifetime, and survivability is the probability of surviving to reproductive age.

I think that #51 could be accomplished by just saving their parents' information, and updating a separate file containing the family tree. If we wanted to know that creature's grandparents, we would look at its parents' name(s), and go to the file and see which names those parents point to in the family tree file. 

Or, instead of using names, we could use some type of globally unique identifier (GUID) since we're still trying to fix issue #52 (also, we'll avoid situations where the same names get repeated). 

For #64, I think speciation should depend on softer barriers. I think I once mentioned the idea of defining a different species as one that has a different number of hidden layers in its brain (partially analogous to how organisms with differing number of chromosomes cannot produce fertile offspring). So different brain dimensions = different species. Is that similar to what you were thinking?

**Comment by kodomazer** [comment]
Everything looks good.

With respect to soft barriers for speciation, if we adopt the brain style I'm working on right now (still need to work out how to do multi-parent generation) then each creature will have an "architecture" for their brain, and if two creatures have architectures that don't match up enough then there is a higher chance of having a defunct baby that doesn't work out.

We should probably have a new issue thread to talk about that.

**Comment by 10carfernandez** [comment]
OK, so I've been working on implementing a method for saving the game, as well as saving creatures and brains. 

As I was outlining what I have done so far and the issues and ideas I came across, I inadvertently rubber ducked some of the issues I came across. 

In short, I think it will be better to use a combination of text and binary files for saving information. The creatures and brains would each be saved as hashmaps in binary files, and only the pointer information (creature and brain IDs, creature dates and locations) would be saved as text files. This will make loading information faster at the expense of taking up a bit more space. 

One problem I saw was that even for small brains, the information builds up VERY quickly. Within about 30 min, the save file for brains managed to accumulate to over 2.5 gb for all of the unique brain IDs that have existed. This isn't a problem for the creature's other information, since it takes up so much less space. So that's why there'd be a separate file for brains - so that we can only add the fittest ones and save space. 

The only time, I think, that we should save ALL brains regardless of fitness is when we want to save the entire game in order to load it later and continue off from where it left off, and during the periodic text file saves, which I suppose are meant to function as save points and can be used to measure fitness over time. 

As far as a family tree goes, I was thinking that we should make a new class called FamilyTree that only stores creature IDs in a structured way, with a method that retrieve other creature info (names, dates, locations, etc.), from the data we're saving, for display purposes.

**Comment by 10carfernandez** [comment]
OK, so since there are a lot of moving parts to what I've been trying to do, I think it might be best to break things down into smaller parts so that I won't have one PR with various somewhat-related bug fixes and features. 

Also, @kodomazer mentioned something about a new type of brain and saving brains. So for now, my first PR will be focused solely on saving and loading an entire game.

I think one thing that might have to change is the folder structure for saving text and image files. My idea is to just keep the folder structure under PIC the same, but have multiple PIC folders (PIC00000, PIC00001, PIC00002, etc.). 

To load games, we might need more than just the extra button we have now. I've changed it so that there are 3 columns of buttons. However, the order/arrangement looks less intuitive/organized now. So I was wondering what you thought about how they can be rearranged. I'll upload a snapshot of what they now look like so you can see it.

I've managed to successfully save and load games now, so now I just have to write code that will allow the user to load files from the game and not overwrite previous files. It'll all make sense once it's done, which will hopefully be sometime tomorrow.

**Comment by 10carfernandez** [comment]
![00000](https://cloud.githubusercontent.com/assets/21129921/22008577/1a8fb062-dc42-11e6-8ae8-51535d6a0987.png)

**Comment by kodomazer** [comment]
We might want to make another window that comes up first with various options, such as save/load path and possibly the config stuff we talked about in #72 and #76 .  This would make it easier for outsiders to come in and use what we've been working on.

In reference to a post I've been in the process of writing, we should be able to manually generate objects and save them as files somehow.  Another thing to think about is that we may want to load an object and then change its location (which pertains to the reason why in the above situation).

**Comment by 10carfernandez** [comment]
Yeah, I was also thinking about having something sort of like Minecraft's start screen where you can start a new simulation or load one, seed options, and other advanced options. 

Before we do that, though, we might want to create a separate file for the UI and/or for the controls. I was thinking about maybe doing that next, after having to make changes to the buttons and write terrible code to make them work in my last PR.

Which brings up a question: I think I remember reading this somewhere, but was there a discussion somewhere about gradually migrating away from the Processing library and making everything straight Java? 

As far as saving objects goes, I was thinking there could be 3 levels:

- At the highest level, save an entire simulation (board) such that you can load it later and pick up from where you last left off. 

- At the next level, be able to save a creature or creatures in a separate file, and load them on any board at any position. That way, people can share "strains" of creatures. 

- At the last level, be able to save brains and load them into creatures. I think one thing this could be useful for is having the option to load a set of "good" starting brains for a simulation so that the user can save time going through randomly generated brains that are not good. 

I was thinking that being able to save and load a simulation right now might help with development too, for when we're trying to catch problems that happen after long periods of time. I know that #52 occurs after running a simulation for a long period of time. So if we can find out when it occurred, we can probably load the file just before the error in a debug session and step right into the values causing the error.
